plugins/log/src/lib.rs

```diff
@@ -142,6 +146,175 @@ impl TimezoneStrategy {
+                    fs::remove_file(&self.path)?;
+                }
+            }
+        }
+        self.open_file()?;
+        Ok(())
+    }
+
+    /// Remove old log files until the number of old log files is equal to the keep_count,
+    /// the current active log file is not included in the keep_count.
+    fn remove_old_files(&self, keep_count: usize) -> Result<(), Error> {
+        let mut files = fs::read_dir(&self.dir)?
+            .filter_map(|entry| {
+                let entry = entry.ok()?;
+                let path = entry.path();
+                let old_file_name = path.file_name()?.to_string_lossy().into_owned();
+                if old_file_name.starts_with(&self.file_name)
+                  // exclude the current active file
+                  && old_file_name != format!("{}.log", self.file_name)
+                {
+                    let date = old_file_name
+                        .strip_prefix(&self.file_name)?
+                        .strip_prefix("_")?
+                        .strip_suffix(".log")?;
+                    Some((path, date.to_string()))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        files.sort_by(|a, b| a.1.cmp(&b.1));
+
+        if files.len() > keep_count {
+            let files_to_remove = files.len() - keep_count;
+            for (old_log_path, _) in files.iter().take(files_to_remove) {
+                fs::remove_file(old_log_path)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn rename_file_to_dated(&self) -> Result<(), Error> {
+        let to = self.dir.join(format!(
+            "{}_{}.log",
+            self.file_name,
+            self.timezone_strategy
+                .get_now()
+                .format(LOG_DATE_FORMAT)
+                .unwrap(),
+        ));
+        if to.is_file() {
+            // designated rotated log file name already exists
+            // highly unlikely but defensively handle anyway by adding .bak to filename
+            let mut to_bak = to.clone();
+            to_bak.set_file_name(format!(
+                "{}.bak",
+                to_bak.file_name().unwrap().to_string_lossy()
+            ));
+            fs::rename(&to, to_bak)?;
+        }
+        fs::rename(&self.path, &to)?;
+        Ok(())
+    }
+}
+
+impl Write for RotatingFile {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.buffer.extend_from_slice(buf);
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        if self.buffer.is_empty() {
+            return Ok(());
+        }
+        if self.inner.is_none() {
+            self.open_file()
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
+        }
+
+        if self.current_size + (self.buffer.len() as u128) > self.max_size {
+            self.rotate()
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
+        }
```

Comment:
If the buffer length is bigger than max_size this will endlessly rotate. Maybe max_size should be a soft limit so that we only rotate after writing something past the max_size? could be bound to a comparison of buffer.lent() and max_size so that this really only happens in those cases.

Reply:
The rotate is called at most one time for one logging call.
The current logic only results in the first log file potentially being rotated before any content is written to it; subsequent log files are written normally.
Perhaps we could check the log file size here: if self.current_size is 0, then we skip rotating check?
```suggestion
        if self.current_size != 0 && self.current_size + (self.buffer.len() as u128) > self.max_size {
            self.rotate()
                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
        }
```

Reply:
righttttt my bad. since log lines larger than max_size should not realistically happen the current code logic is fine but i wouldn't mind the 0 check so we really never end up with empty files - your call though.